resolver/resolver.cc

```diff
@@ -1311,8 +1533,57 @@ class ResolveTypeMembersWalk {
         return true;
     }
 
+    // Returns `true` if `asgn` is a field declaration.
+    bool handleFieldDeclaration(core::Context ctx, ast::Assign &asgn) {
+        auto *uid = ast::cast_tree<ast::UnresolvedIdent>(asgn.lhs);
+        if (uid == nullptr) {
+            return false;
+        }
+
+        if (uid->kind != ast::UnresolvedIdent::Kind::Instance && uid->kind != ast::UnresolvedIdent::Kind::Class) {
+            return false;
+        }
+
+        auto *recur = &asgn.rhs;
+        while (auto *outer = ast::cast_tree<ast::InsSeq>(*recur)) {
+            recur = &outer->expr;
+        }
+
+        auto *cast = ast::cast_tree<ast::Cast>(*recur);
+        if (cast == nullptr) {
+            return false;
+        } else if (cast->cast != core::Names::let()) {
+            if (auto e = ctx.beginError(cast->loc, core::errors::Resolver::ConstantAssertType)) {
+                e.setHeader("Use `{}` to specify the type of constants", "T.let");
+            }
+        }
+
+        ResolveFieldItem &job = todoResolveFieldItems_.emplace_back();
+        job.file = ctx.file;
+        job.owner = ctx.owner;
+        job.cast = cast;
+        job.ident = uid;
+        job.atTopLevel = nestedBlockCounts.back() == 0;
```

Comment:
Why did we chose this method? If someone adds a new field, clang won't catch all the places where you need to now initialize that field.

Versus

```cpp
todoResolveFieldItems_.emplace_back(ResolveFieldItem{
  ctx.file,
  ctx.owner,
  cast,
  uid,
  nestedBlockCounts.back() == 0,
});
```

Reply:
It's self-documenting; the way you have it doesn't document the field names, and clang won't complain if I miss one (it'll default initialize it). But I have no strong preference here and can do it that way.

Reply:
> It's self-documenting

definitely agree that this is nice



> clang won't complain if I miss one (it'll default initialize it)

this is the one that worries me

once we upgrade to C++20 we will be able to use designated initializers to make initializer lists like that self documenting:

https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers

since you wrote it the way you did i'm happy to leave it for now, but i might make another pass over in our codebase at some point.

Reply:
I fixed it in https://github.com/sorbet/sorbet/pull/3771